src/sentry/models/apiapplication.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import absolute_import, print_function
+
+import petname
+
+from django.db import models
+from django.utils import timezone
+from django.utils.translation import ugettext_lazy as _
+from uuid import uuid4
+
+from sentry.db.models import (
+    Model, BaseManager, BoundedPositiveIntegerField, EncryptedTextField,
+    FlexibleForeignKey, sane_repr
+)
+
+
+class ApiApplicationStatus(object):
+    active = 0
+    inactive = 1
+    pending_deletion = 2
+    deletion_in_progress = 3
+
+
+class ApiApplication(Model):
+    __core__ = True
+
+    client_id = models.CharField(
+        max_length=64, unique=True,
+        default=lambda: ApiApplication.generate_token())
+    client_secret = EncryptedTextField(
+        default=lambda: ApiApplication.generate_token())
+    owner = FlexibleForeignKey('sentry.User')
+    name = models.CharField(
+        max_length=64, blank=True,
+        default=lambda: petname.Generate(2, ' ').title())
+    status = BoundedPositiveIntegerField(default=0, choices=(
+        (ApiApplicationStatus.active, _('Active')),
+        (ApiApplicationStatus.inactive, _('Inactive')),
+    ), db_index=True)
+    allowed_origins = models.TextField(blank=True, null=True)
+    redirect_uris = models.TextField()
+
+    homepage_url = models.URLField(null=True)
+    privacy_url = models.URLField(null=True)
+    terms_url = models.URLField(null=True)
+
+    date_added = models.DateTimeField(default=timezone.now)
+
+    objects = BaseManager(cache_fields=(
+        'client_id',
+    ))
+
+    class Meta:
+        app_label = 'sentry'
+        db_table = 'sentry_apiapplication'
+
+    __repr__ = sane_repr('name', 'owner_id')
+
+    def __unicode__(self):
+        return self.name
+
+    @classmethod
+    def generate_token(cls):
+        return uuid4().hex + uuid4().hex
+
+    @property
+    def is_active(self):
+        return self.status == ApiApplicationStatus.active
+
+    def is_allowed_response_type(self, value):
+        return value in ('code', 'token')
+
+    def is_valid_redirect_uri(self, value):
+        for ruri in self.redirect_uris.split('\n'):
```

Comment:
I think this might be an issue. If someone has `http://example.com` set as valid redirect uri someone could use `http://example.com.mydomain.com` and get a valid redirect target.